src/lib/transition_handler/processor.ml

```diff
@@ -0,0 +1,29 @@
+open Protocols.Coda_pow
+
+module Make (Inputs : Inputs_intf.S) : Transaction_handler_processor_intf = struct
+  open Inputs
+
+  (* TODO: implement drop old *)
+  let run ~logger ~valid_transition_reader ~catchup_job_writer ~catchup_breadcrumbs_reader frontier =
+    let logger = Logger.parent logger "Transition_handler.Catchup" in
+    let catchup_monitor = Catchup_monitor.create ~catchup_job_writer in
+    let reader = Linear_pipe.Priority_reader.create catchup_transitions_reader valid_transition_reader in
+    
+    Linear_pipe.iter reader ~f:(function
+    | `Catchup_transitions [] -> Logger.error logger "read empty catchup transitions"
+    | `Catchup_transitions ((root_breadcrumb :: _) as breadcrumbs) ->
+        (match Transition_frontier.lookup frontier (breadcrumb_parent_hash root_breadcrumb) with
+        | None ->
+            Logger.error logger "read catchup transitions which did not fit into frontier"
+        | Some parent ->
+            List.fold breadcrumbs ~init:parent ~f(fun parent breadcrumb -> 
+                Transition_frontier.add frontier ~parent ~breadcrumb))
```

Comment:
breadcrumbs need to know about their successors (this has to be all hooked in) because we can only afford one `add_exn` call here. It's too expensive to do them in a fold.

Reply:
We have to do this. This is only for the transitions returned by catchup. However, there is a disconnect here between what the frontier code does. We should talk about this. The idea here was that the breadcrumb would already have been calculated, so it can be added without deriving the new mask (which would have been done in catchup, and is the expensive part of adding to the tree).

We have to block and add those at some point. We should just add a second interface to transition frontier that splits up the creation of a breadcrumb and adding it to the tree.